src/iterative/KOSMAPOSL/KOSMAPOSLReconstruction.cxx

```diff
@@ -0,0 +1,1119 @@
+
+                        const int min_x = (*anatomical_sptr)[z][y].get_min_index();
+                        const int max_x = (*anatomical_sptr)[z][y].get_max_index();
+
+
+       for (int x=min_x;x<= max_x;x++)
+                          {
+                            const int min_dx = max(distance[0][0].get_min_index(), min_x-x);
+                            const int max_dx = min(distance[0][0].get_max_index(), max_x-x);
+
+                            for (int dz=min_dz;dz<=max_dz;++dz)
+                              for (int dy=min_dy;dy<=max_dy;++dy)
+                                for (int dx=min_dx;dx<=max_dx;++dx)
+                                  {
+                                    if (get_hybrid()){
+
+                                        if(current_estimate[z][y][x]==0){
+                                             continue;
+
+                                        }
+                                        else{
+
+                                        kPET=exp(-square((current_estimate[z][y][x]-current_estimate[z+dz][y+dy][x+dx])/current_estimate[z][y][x]*get_sigma_p())/2)*
+                                             exp(-square(distance[dz][dy][dx]/grid_spacing.x ())/(2*square(get_sigma_dp())));
+                                        }
+
+                            }
+                            else{
+                                kPET=1;
+
+                            }  // the following "pnkernel" is the normalisation of the kernel
+                                    kanatomical=exp(-square(((*anatomical_sptr)[z][y][x]-(*anatomical_sptr)[z+dz][y+dy][x+dx])/anatomical_sd*sigma_m)/2)*
+                                                 exp(-square(distance[dz][dy][dx]/grid_spacing.x ())/(2*square(sigma_dm)));
+
+                                    pnkernel+=kPET*kanatomical;
+
+                                    kImage[z][y][x] += kanatomical*kPET*Image[z+dz][y+dy][x+dx];//
+
+                                   }
+                                    if(current_estimate[z][y][x]==0){
+                                         continue;}
+                                     kImage[z][y][x]= kImage[z][y][x]/pnkernel;
+                                     pnkernel=0;
+                       }
+                    }
+              }
+
+}
+
+template<typename TargetT>
+void KOSMAPOSLReconstruction<TargetT>::compute_kernelised_image(
+                         TargetT& kImage,
+                         const TargetT& Image,
+                         const TargetT& current_estimate)
+{
+
+    if(this->num_non_zero_feat==1){
+        compact_compute_kernelised_image (kImage, Image, current_estimate);
+                                    }
+    else{
+      full_compute_kernelised_image (kImage, Image,
+                               current_estimate);
+}
+}
+
+template <typename TargetT>
+void 
+KOSMAPOSLReconstruction<TargetT>::
+update_estimate(TargetT &current_alpha_coefficent_image)
```

Comment:
I think the point of the refactoring of OSMAPOSLReconstruction is so that we wouldn't have to modify update_estimate. Did you run into trouble with the other approach?